train.py

```diff
@@ -1,90 +1,129 @@
-
-        for i, data in enumerate(loader):
-            with Timer("Elapsed time in update " + str(i) + ": %f"):
-                iter_start_time = time.time()
-                if total_steps % opt.train.print_freq == 0:
-                    t_data = iter_start_time - iter_data_time
-                total_steps += opt.data.loaders.batch_size
-                epoch_iter += opt.data.loaders.batch_size
-
-                model.set_input(Dict(data))
-                model.optimize_parameters()
-
-                if total_steps % opt.val.save_im_freq == 0:
-                    model.save_test_images(test_display_images, total_steps)
-
-                if total_steps % opt.train.save_freq == 0:
-                    print(
-                        "saving the latest model (epoch %d, total_steps %d)"
-                        % (epoch, total_steps)
-                    )
-                    save_suffix = "iter_%d" % total_steps
-                    model.save_networks(save_suffix)
-
-                iter_data_time = time.time()
-
-        """
-        if epoch % opt.save_epoch_freq == 0:
-            print("saving the model at the end of epoch %d, iters %d" % (epoch, total_steps))
-            model.save_networks("latest")
-            model.save_networks(epoch)
-
-        print(
-            "End of epoch %d / %d \t Time Taken: %d sec"
-            % (epoch, opt.niter + opt.niter_decay, time.time() - epoch_start_time)
-        )
-        model.update_learning_rate()
-        """
+    times = deque([0], maxlen=100)
+    model_times = deque([0], maxlen=100)
+    batch_size = opts.data.loaders.batch_size
+    checkpoint_directory, image_directory = prepare_sub_folder(opts.train.output_dir)
+    tpe = opts.train.tests_per_epoch
+    test_idx = [i * len(train_loader) // tpe for i in range(tpe)]
+    test_idx[-1] = len(train_loader) - 1
+    test_idx = set(test_idx)
+
+    # ---------------------------
+    # -----  Training Loop  -----
+    # ---------------------------
+    s = "Starting training for {} epochs of {} updates with batch size {}, "
+    s += "{} test inferences per epoch."
+    print(s.format(opts.train.epochs, len(train_loader), batch_size, tpe))
+
+    for epoch in range(opts.train.epochs):
+        print(f"Epoch {epoch}: ")
+        comet_exp.log_metric("epoch", epoch, step=total_steps)
+        for i, data in enumerate(train_loader):
+            times.append(time())
+            total_steps += batch_size
+
+            model.set_input(Dict(data))
+            model.optimize_parameters(total_steps)
+
+            model_times.append(time() - times[-1])
+            if total_steps // batch_size % 100 == 0:
+                avg = avg_duration(times, batch_size)
+                mod_times = np.mean(model_times) / batch_size
+                comet_exp.log_metric("sample_time", avg, step=total_steps)
+                comet_exp.log_metric("model_time", mod_times, step=total_steps)
+            if i in test_idx or total_steps == batch_size:
```

Comment:
Does that mean the images are logged at each epoch ? Isn't that a bit much ?

Reply:
This means that it logs images once after the first update and then tests_per_epoch times per epoch, which I defaulted to 4

You'd rather log more rarely? Less than 1 once per epoch? @51N84D thoughts?

Reply:
I think I'm used to loggig more rarely, just because you end up with a lot of images at the beginning that you don't really need/look at , but it doesn't take that much memory I guess so I get your point.